dmlex-v1.0/specification/core/specification.xml

```diff
@@ -64,7 +64,60 @@
         </listitem>
      </itemizedlist>
 	</para>
-  </simplesect>
+	  <section id="frag_iri">
+		  <title>Fragment IRIs</title>
+		  <para>DMLex provides a recommended method for addressing DMLex objects present on-line, useful for linking (cf. <xref linkend="linking"/>) and general interoperability. Implementing this method is not <glossterm>required</glossterm> for conformance.</para>
+
+		  <para>Every fragment <glossterm>should</glossterm> be assigned a unique IRI (Internationalized Resource Identifier [<link linkend="bib_rfc3987">RFC 3987</link>]), composed of <literal>lexicographicResource.uri</literal> and a sequence of identifiers that uniquely determines the path in the DMLex tree structure. The IRI of the root object <literal>lexicographicResource</literal> is the value of its attribute <literal>lexicographicResource.uri</literal>, converted to IRI according to the algorithm specified in [<link linkend="bib_rfc3987">RFC 3987</link>], if needed. The IRIs of its direct children are constructed as follows:</para>
+
+		<para><literal>lexicographicResource.uri/objectTypeName/objectID</literal></para>
```

Comment:
Shouldn't this be `lexicographicResource.uri#objectTypeName/objectID`?

Reply:
This PR claims to create _fragments_ but does not. _Fragments_ are the portion of an (HTTP) URI that occur after the `#` symbol. This is an important distinction as, for example this URL `http://www.example.com/lexicon/lexicographicResource/entry/cat` refers to a document that describes only the entry cat. In contrast `http://www.example.com/lexicon#lexicographic/entry/cat` refers to the identified section of the document `http://www.example.com/lexicon`.

Reply:
I think this is just a terminological misunderstanding. "Fragment" in fragment identification does not refer to URI fragment. It's merely a fragment in the sense of part of the data.

Reply:
First, the term "fragment" is pretty widely understood and I wouldn't redefine it. Secondly, I think you *do* want fragments in this sense as otherwise it is very challenging to create URIs that resolve, and this would be a big technical issue.

Reply:
- well, I was placing my content within the already existing section **Fragment identification**, so I guess the meaning of the word "fragment" was already (re)defined and it has the sense of (partial) DMLex objects. I could rename 3.2.1 Fragment IRIs to Object IRIs to avoid confusion but I don't want to touch 3.2 now as it was introduced before me (unless there is a consensus that it should be renamed)
- I don't think we want URI fragments, don't understand how is it challenging, can you explain? My view on URI fragments is that they are anchors within the response of the previous part of the URI -- e.g. you download a web page based on the URI without the fragment and scroll down to the anchor defined by the fragment. Also the RFC says

> The fragment's format and resolution is therefore dependent on the media type of a potentially retrieved representation, even though such a retrieval is only performed if the URI is dereferenced.

I don't think this matches our semantics, we want to identify the objects directly, not as anchors within the whole lexicographic resource. (But I don't think it's something extremely important, will not fight against fragments if more of you think it's better.)

Reply:
The problem is that this does not work as a *URL* and I assume that we want our IRIs to be interpretable as HTTP URLs in most cases.

An HTTP URL consists of a *path* and a *fragment*, and the server requests a single document at the *path* and uses the *fragment* to identify the location in that document.

If we assign a path-based identifier to, for example, a sense, then that implies that the server should return a document that represents that sense when queried with this path. However, our XML and JSON serializations do not allow a `sense` to be a root element, so there is no document that can be returned. Instead, the sense is a section of the overall lexicon document, which fits with the semantics of a fragment.

Reply:
I do not think we ever discussed that we would want the IRIs to be usable as URLs so this is a far reaching implicit assumption that is false at this moment.

But even if we agree that we want this, I strongly object against adhering any such notion of document in the sense of what HTTP or similar protocols might have defined ages ago and noone actually follows. The IRIs are strictly to be understood as links within DMLex-internal addressing mechanisms, and under no circumstances as addresses giving any expectations as to what they should return, particularly not within the framework of one arbitrary protocol such as HTTP.

Btw, even for HTTP, I think it is completely fine for implementers to return whole documents, or any portions of them -- DMLex is not designed to be a round-trip mechanism, this is simply out of the scope.

Reply:
> I do not think we ever discussed that we would want the IRIs to be usable as URLs so this is a far reaching implicit assumption that is false at this moment.

URIs starting with `http` are HTTP URLs. The examples you have given are HTTP URLs so the assumption is pretty clear.

> But even if we agree that we want this, I strongly object against adhering any such notion of document in the sense of what HTTP or similar protocols might have defined ages ago and noone actually follows. 

HTTP is a foundational standard of the web, I don't understand why you think no-one follows it

> The IRIs are strictly to be understood as links within DMLex-internal addressing mechanisms, and under no circumstances as addresses giving any expectations as to what they should return, particularly not within the framework of one arbitrary protocol such as HTTP.

I am not sure I agree on the need an internal addressing mechanism, but if we do introduce a mechanism like this (once we have ironed out the bugs), it should not be a mechanism that looks like HTTP but does not function like HTTP. Creating our own rules that contradict one of the most widely deployed standards is only likely to lead to confusion and challenges in implementation.

> Btw, even for HTTP, I think it is completely fine for implementers to return whole documents, or any portions of them -- DMLex is not designed to be a round-trip mechanism, this is simply out of the scope.

Returning the whole lexicon document for all URIs is the behaviour we get for free if we implement this using `#` fragments. Otherwise we put a lot of technical questions to implementers, such as whose job it is to validate these identifiers and how can this be implemented on widely-used servers (Apache, nginx, etc.). 

I don't understand your resistance here. This comment is about changing *one* character in a URI to make it conformant with widely-used standards.

Reply:
(My comments about IRIs/URLs are mainly about resolvability: we do not enforce that, so in that sense that might not be a valid/usable URL (and it will frequently be the case it is not usable) --- but that's a minor point.)

The core of my objection is though the fact that the whole dictionary is not to be seen as one HTTP document -- this is completely up to the implementation what a document is in the context of HTTP.

So: *Returning the whole lexicon document for all URIs is the behaviour we get for free if we implement this using # fragments* is very much an unwanted behaviour. Depending on the context, I want to be able make *different* **HTTP fragments** over the same **DMLex fragment**. Such as that you have links between entries or senses, but you want to navigate the user to a particular example or some other part within an HTML page -- making anything after the `lexicographicResource` being a URL fragment makes it impossible to anchor anything within a particular entry.

Reply:
> The core of my objection is though the fact that the whole dictionary is not to be seen as one HTTP document -- this is completely up to the implementation what a document is in the context of HTTP.

Our standard says that the lexicon (or entry) is a single document in XML and JSON serializations. 

> So: Returning the whole lexicon document for all URIs is the behaviour we get for free if we implement this using # fragments is very much an unwanted behaviour. Depending on the context, I want to be able make different HTTP fragments over the same DMLex fragment. Such as that you have links between entries or senses, but you want to navigate the user to a particular example or some other part within an HTML page -- making anything after the lexicographicResource being a URL fragment makes it impossible to anchor anything within a particular entry.

I am not really sure I understand... "navigating to a particular [element] within an HTML page" is the use case of fragments. A particular application could easily further extend this fragment scheme if they wish so there is no challenge with adding extra fragments to the "DMLex fragments", we are simply defining one mechanism within a DMLex document.

Reply:
I don't think the argument is about following or not following HTTP conventions -- we all want to follow that. I think the core disagreement is about this:

> Returning the whole lexicon document for all URIs is the behaviour we get for free

That's right -- but I think this behaviour does not play well with DMLex principles -- I always felt like its nature are interlinked objects, with lexicographicResource being just one of them and nothing special. Using HTTP fragments would make it _very_ special, and would (kind of) enforce downloading the whole lexicographic resource whenever asking for a single entry, sense, or even example. I don't like that. (But I am still new to these discussions, and yes, it's just one character, so do say and I will do as you say ;-))

Reply:
> > The core of my objection is though the fact that the whole dictionary is not to be seen as one HTTP document -- this is completely up to the implementation what a document is in the context of HTTP.
> 
> Our standard says that the lexicon (or entry) is a single document in XML and JSON serializations.

Yes, but that's a completely different thing and it is relevant only for those two particular serializations. The addressing mechanism is not serialization-specific, so this is not relevant.

> > So: Returning the whole lexicon document for all URIs is the behaviour we get for free if we implement this using # fragments is very much an unwanted behaviour. Depending on the context, I want to be able make different HTTP fragments over the same DMLex fragment. Such as that you have links between entries or senses, but you want to navigate the user to a particular example or some other part within an HTML page -- making anything after the lexicographicResource being a URL fragment makes it impossible to anchor anything within a particular entry.
> 
> I am not really sure I understand... "navigating to a particular [element] within an HTML page" is the use case of fragments. A particular application could easily further extend this fragment scheme if they wish so there is no challenge with adding extra fragments to the "DMLex fragments", we are simply defining one mechanism within a DMLex document.

That's not true -- the RFC for URI (https://datatracker.ietf.org/doc/html/rfc3986) is very clear that the fragment may not contain a hash sign.

The bottom line is: 

* the addressing mechanism is not serialization specific, there is no concept of a document 
* if (rarely) the URLs would be resolvable, we want to make it possible to anchor to arbitrary response parts, therefore we do not want to use the hash sign anywhere.